src/app/hooks/useSafariCompositionTaggingForKeyDown.ts

```diff
@@ -0,0 +1,33 @@
+import { useEffect } from 'react';
+
+const actuallyComposingTag = Symbol("event is actually composing")
+
+export function isTaggedAsComposing(x: object): boolean {
+  return actuallyComposingTag in x
+}
+
+export function useSafariCompositionTaggingForKeyDown(target: Window, {compositionEndThreshold = 500}: {compositionEndThreshold?: 500} = {}) {
+  useEffect(() => {
+    let compositionJustEndedAt: number | null = null
+
+    function recordCompositionEnd(evt: CompositionEvent) {
+      compositionJustEndedAt = evt.timeStamp
+    }
+
+    function interceptAndTagKeyDown(evt: KeyboardEvent) {
+      if (compositionJustEndedAt !== null
+        && evt.keyCode === 229
+        && (evt.timeStamp - compositionJustEndedAt) < compositionEndThreshold) {
+        Object.assign(evt, { [actuallyComposingTag]: true })
```

Comment:
`actuallyComposingTag` is the variable outside the scope of this hook and is side effect, the `actuallyComposingTag` variable is a good candidate for the /app/state as atom if you want to retain state and available throughout app.